File: solutions/fsharp/bird-watcher/1/BirdWatcher.fs
module BirdWatcher

let lastWeek: int[] = [| 0; 2; 5; 3; 7; 8; 4 |]

let yesterday (counts: int[]) : int = counts[counts.Length - 2]

let total (counts: int[]) : int = counts |> Array.sum

let dayWithoutBirds (counts: int[]) : bool =
    Array.tryFind (fun num -> num = 0) counts <> None

let incrementTodaysCount (counts: int[]) : int[] =
    Array.append (Array.take (counts.Length - 1) counts) ([| Array.last counts + 1 |])

let unusualWeek (counts: int[]) : bool =
    let countsWithIdx = counts |> Array.zip [| 1 .. counts.Length |]

    let evenNoBirds =
        countsWithIdx |> Array.forall (fun (idx, count) -> idx % 2 = 1 || count = 0)

    let evenTenBirds =
        countsWithIdx |> Array.forall (fun (idx, count) -> idx % 2 = 1 || count = 10)

    let oddFiveBirds =
        countsWithIdx |> Array.forall (fun (idx, count) -> idx % 2 = 0 || count = 5)

    evenNoBirds || evenTenBirds || oddFiveBirds

File: solutions/csharp/bird-watcher/1/BirdWatcher.cs
using System;
using System.Linq;

class BirdCount
{
    private int[] birdsPerDay;

    public BirdCount(int[] birdsPerDay)
    {
        this.birdsPerDay = birdsPerDay;
    }

    public static int[] LastWeek() => new[] { 0, 2, 5, 3, 7, 8, 4 };

    public int Today() => birdsPerDay.Last();

    public void IncrementTodaysCount() => birdsPerDay[birdsPerDay.Length - 1]++;

    public bool HasDayWithoutBirds() => birdsPerDay.Any(day => day == 0);

    public int CountForFirstDays(int numberOfDays) => birdsPerDay[..numberOfDays].Sum();

    public int BusyDays() => birdsPerDay.Where(day => day >= 5).Count();
}
